lib/mas/firm_indexer.rb

```diff
@@ -0,0 +1,32 @@
+module FirmIndexer
+  class << self
+    def index_firm(firm)
+      if !firm.destroyed? && firm.publishable?
+        store_firm(firm)
+      else
+        delete_firm(firm)
+      end
+    end
+
+    def handle_firm_changed(firm)
+      index_firm(firm)
+    end
+
+    def handle_aggregate_changed(aggregate)
+      index_firm(aggregate.firm) if firm_exists?(aggregate.firm)
```

Comment:
This is very clear is it. But there's a difference in context.

In `index_firm` if the firm is "destroyed" we want to remove it from the index.

In `handle_aggregate_changed` if the firm is destroyed/doesn't exist we want to do nothing.

If we move the guard into `index_firm` we would need to work out whether it was coming from `handle_aggregate_changed`.

Maybe `firm_exists?` is a poor name. What it is being used to express is, "are we in the middle of a cascade delete?"